src/main/java/com/microsoft/azure/relay/HybridConnectionListener.java

```diff
@@ -277,50 +276,46 @@ public HybridConnectionListener(String connectionString, String path) throws URI
 	 *         disconnected with the cloud service
 	 */
+	@SuppressWarnings("unchecked")
 	public CompletableFuture<Void> closeAsync(Duration timeout) {
-		try {
-			CompletableFutureUtil.timedRunAsync(timeout, () -> {
-					List<HybridConnectionChannel> clients;
-					synchronized (this.thisLock) {
-						if (this.closeCalled) {
-							return;
-						}
-	
-						RelayLogger.logEvent("closing", this);
-						this.closeCalled = true;
-	
-						// If the input queue is empty this completes all pending waiters with null and
-						// prevents any new items being added to the input queue.
-						this.connectionInputQueue.shutdown();
-	
-						// Close any unaccepted rendezvous. DequeueAsync won't block since we've called
-						// connectionInputQueue.Shutdown().
-						clients = new ArrayList<HybridConnectionChannel>(this.connectionInputQueue.getPendingCount());
-						HybridConnectionChannel webSocket;
-						do {
-							webSocket = this.connectionInputQueue.dequeueAsync().join();
-	
-							if (webSocket != null) {
-								clients.add(webSocket);
-							}
-						} while (webSocket != null);
-	
-						this.isOnline = false;
-					}
-	
-					clients.forEach(client -> client.closeAsync(
-							new CloseReason(CloseReason.CloseCodes.NORMAL_CLOSURE, "Client closing the socket normally")));
-	
-					RelayLogger.logEvent("closed", this);
-				},
-				EXECUTOR).join();
-		}
-		catch (Exception e) {
-			throw RelayLogger.throwingException(e, this);
-		} finally {
+		return CompletableFutureUtil.timedSupplyAsync(timeout, () -> {
+			synchronized (this.thisLock) {
+				if (this.closeCalled) {
+					return CompletableFuture.completedFuture(null);
+				}
+
+				RelayLogger.logEvent("closing", this);
+				this.closeCalled = true;
+
+				// If the input queue is empty this completes all pending waiters with null and
+				// prevents any new items being added to the input queue.
+				this.connectionInputQueue.shutdown();
+
+				// Close any unaccepted rendezvous. DequeueAsync won't block since we've called
+				// connectionInputQueue.Shutdown().
+				CompletableFuture<?>[] closeTasks = new CompletableFuture<?>[this.connectionInputQueue.getPendingCount()];
+				for (int i = 0; i < this.connectionInputQueue.getPendingCount(); i++) {
+					closeTasks[i] = this.connectionInputQueue.dequeueAsync().thenAccept(connection -> {
+						connection.closeAsync(
+								new CloseReason(CloseReason.CloseCodes.NORMAL_CLOSURE, "Client closing the socket normally"));
+					});
+				}
+				return closeTasks;
+			}
+		}, EXECUTOR)
+		.thenCompose(closeTasks -> {
+			return CompletableFuture.allOf((CompletableFuture<Void>[]) closeTasks).thenRun(() -> {
+				this.isOnline = false;
+				RelayLogger.logEvent("closed", this);
+			});
+		})
+		.whenComplete((nullResult, ex) -> {
 			this.connectionInputQueue.dispose();
-		}
-		return this.controlConnection.closeAsync(null);
+			if (ex != null) {
+				throw RelayLogger.throwingException(ex, this);
```

Comment:
>throw RelayLogger.throwingException(ex, this); [](start = 4, length = 46)

Does this throw cause the controlConnection to never get closed?

Reply:
In case of an exception, most likely an IO exception, the ControlConnection probably won't be in a state to be closed normally anyways right?

---
In reply to: [257775403](https://github.com/Azure/azure-relay-java/pull/14#discussion_r257775403) [](ancestors = 257775403)

Reply:
See what you think of this PR:
[https://github.com/Azure/azure-relay-java/pull/15](https://github.com/Azure/azure-relay-java/pull/15)

---
In reply to: [258610208](https://github.com/Azure/azure-relay-java/pull/14#discussion_r258610208) [](ancestors = 258610208,257775403)